capi-configuration/3_upgrades.md

```diff
@@ -0,0 +1,52 @@
+# Handling workload cluster upgrades
+
+## Problem Statement
+
+We want to ...
+1. Make all changes transparent for the user before application
+2. Continue supporting the current workflow of upgrading a cluster (Through GS staff, customers and automation)
+3. Integrate the process into the proposed gitops-management
+4. Utilize the managed-configmap approach to ensure clusters adhere to the definition in their release at all times
+5. Ensure that an upgrade to a cluster is decoupled from upgrades of other components
+6. Allow easy integration into other pipelines and formats users might want to utilize for cluster management.
+
+## Triggering an Upgrade
+
+Our current method to trigger an upgrade is to change the release-version label on the `Cluster` resource. This can be done by GS staff, customers and automation through various interfaces.
+This change will trigger the actual upgrade by letting a newer set of components manage the cluster.
+However, when introducing the [new structure for releases](0_capi-releases), this process will be different. The release will no longer be tied to the components.
+Instead, each release version will relate to a set of configmaps, defining the defaults and basic structure for the cluster resources.
+So, triggering an upgrade in the future will mean to request the change of certain values within the CR itself. This is a breaking change. (changing the release version label alone will no longer trigger an upgrade)
+
+## Upgrade Mechanism
```

Comment:
I think this should explain the process of "someone pressed a button" we had discussed.

Additionally we should reference the RFC by fer for an interface:
https://github.com/giantswarm/rfc/tree/main/automatic-cluster-upgrades

Reply:
Hmm can you be more specific about the "someone pressed a button" we had discussed?
I was thinking to add more to the `Client` section. The automatic cluster upgrades would fit there as well.

Reply:
IMHO ideally the upgrade mechanism is always the same on a lower level, different clients would just enable it to be used in different ways.

Example:
- assumed kgs (gitops) approach would be as described below in the slides (the exact approach is not directly impactful in this context)
- automated upgrade (for non-gitops clusters) would do exactly the same, just that the whole procedure would be done by an operator
- happa/button (for non-gitops clusters) basically could re-use the same mechanics as the automated upgrade but without scheduling (or even with scheduling as someone could schedule an upgrade for tomorrow midnight)

Looking at the example, basically we can differentiate between two upgrade modes:
- gitops mode where the source of truth is in git (interfaces need to be either outputting files or doing actual PRs to git repos, automated upgrade can only be done with automation that has access to the users' git)
- non-gitops mode where the source of truth is in the MC (ideally there's an operator that can mutate resources towards an upgrade, that can then be triggered by different interfaces, i.e. CLI, Happa, auto-upgrade/scheduler:)

Reply:
I agree that the upgrades should always be the same on a lower level, but that seems to be at odds with our current approach to implement it to `kubectl-gs` directly.
In this case it would be indeed very similar between gitops and non-gitops because we basically generate the proposed changes, review them and then either apply them directly or to a repo.
We can probably replicate the same in happa at least somewhat. (Even if at the beginning we just have to make things read-only for gitops-managed clusters)

In case of the scheduled approach, it's a bit different. We could do the exact same workflow also automated through an operator. For gitops-managed clusters this has to lead to an automated commit to the managing repo. For other clusters it would lead to an automated apply.
But then we would lose the ability to review which IMO is not good. 
So what I'm thinking right now is that rather than the review and actual upgrade being one and the same, it would be good to define upgrades in their own resource. (I believe fer suggested something similar) and have only one component execute the upgrades based on this resource. (How the execution looks like would then differ between gitops and non-gitops as you described)
So using kubectl-gs we would still get to review or apply the upgrade. Reviewing would be as described here. If we apply it, an upgrade resource would be created. Furthermore we could define the upgrade time and include scheduling directly.
Happa would do the same.

Reply:
just saw this here, copying my comment from Slack:
one thing that’s on my mind reading the “scheduling everywhere” is that in a gitops approach we'd be basically scheduling a PR, but the intention of an upgrade schedule is to run the upgrade at that time, not the PR which then again needs reviewing. In this context I’m not sure scheduling outside of the actual gitops tool makes sense when using gitops. Maybe in a gitops world the PR would even be created right when the release is GA and then the user “schedules” the upgrade through git and/or gitops tool automation (e.g. bot driven auto-merge, workflows,…)

Reply:
Yeah I absolutely agree.
Also paraphrasing from slack, I don't think we need to worry much at all about the scheduling and gitops.
Rather than that it is more that we should make sure all upgrades, scheduled or not have the same source. So if as suggested we use `kubectl-gs` to create and review upgrades which can be used in gitops, then we should also be able to get a scheduled upgrade using the same command/basic mechanism.